### PR TITLE
[igemm] this kernel can support resnet50 first layer wrw

### PR DESCRIPTION
**Pull request by asleepzzz**
resnet50 first layer fp16 is 8TF,
this kernel is 28TF

### CODE REVIEW

**Comment by asroy** [comment]
@asleepzzz Please watch the CI and make sure this PR pass

File: src/include/miopen/solver.hpp [hunk 1]
```diff
@@ -1981,6 +1981,83 @@ struct ConvHipImplicitGemmWrwV4R4Xdlops : SolverBase<ConvolutionContext>
...
+    int GemmNPerBlock;
+    int GemmKPerBlock;
+    int GemmMPerWave;
+    int GemmNPerWave;
+    int GemmKPack;
+    int GemmMFactor;
+    int GemmNFactor;
+    int GemmKFactor;
+    bool GemmAThreadCopyMoreGemmK;
+    bool GemmBThreadCopyMoreGemmKPack;
```
Comment (ltqin) [inline src/include/miopen/solver.hpp:1998]: change it to GemmBThreadCopyMoreGemmK

File: src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_backward_weights_implicit_gemm_v4r4_xdlops_nchw_kcyx_nkhw_padded_gemm.hpp [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            make_tuple(Merge<Sequence<G, GemmKBlocks>>{},
+                       PassThrough<GemmM - GemmMPad>{},
+                       Pad<Sequence<GemmN - GemmNPad>, Sequence<0>, Sequence<GemmNPad>>{}),
+            make_tuple(Sequence<1, 0>{}, Sequence<2>{}, Sequence<3>{}),
+            make_tuple(Sequence<0>{}, Sequence<1>{}, Sequence<2>{}));
+
+        constexpr auto wei_gemmg_gemmm_gemmn_global_desc = transform_tensor_descriptor(
+            wei_gemmg_gemmm_gemmn_padding_global_desc,
+            make_tuple(PassThrough<GemmG * GemmKBlocks>{},
+                       Pad<Sequence<GemmM - GemmMPad>, Sequence<0>, Sequence<GemmMPad>>{},
```
Comment (ltqin) [inline src/kernels/composable_kernel/include/kernel_algorithm/gridwise_convolution_backward_weights_implicit_gemm_v4r4_xdlops_nchw_kcyx_nkhw_padded_gemm.hpp:234]: why is  pading GemmN and GemmM in two transformation?

File: src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp [hunk 3]
```diff
@@ -0,0 +1,1106 @@
...
+    int GemmNPerBlock_,
+    int GemmKPerBlock_,
+    int GemmMPerWave_,
+    int GemmNPerWave_,
+    int GemmKPack_,
+    int GemmMFactor_,
+    int GemmNFactor_,
+    int GemmKFactor_,
+    bool GemmAThreadCopyMoreGemmK_,
+    bool GemmBThreadCopyMoreGemmKPack_)
```
Comment (ltqin) [inline src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp:60]: Chang the name

File: src/include/miopen/solver.hpp (hunk 1 above)
Comment (asroy) [reply src/include/miopen/solver.hpp:1998]: @asleepzzz Please check the non-padded bwd-wrw kernel, this part of code has been updated in that kernel.

File: src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp [hunk 4]
```diff
@@ -0,0 +1,1106 @@
...
+    {
+        // list performance parameters in reverse order, in order for tuning to iterate over the
+        // range in normal order
+        if(!NextFlag<false, true>(GemmBThreadCopyMoreGemmKPack))
+            break;
+        if(!NextFlag<false, false>(GemmAThreadCopyMoreGemmK))
+            break;
+        if(!NextTwoPower<1, 8>(GemmKPack))
+            break;
+        if(!NextTwoPower<32, 128>(GemmNPerWave))
```
Comment (asroy) [inline src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp:105]: The tuning range for non-padded bwd-wrw has been enlarged, please check that kernel and also update this one

File: src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp [hunk 5]
```diff
@@ -0,0 +1,1106 @@
...
+#include <miopen/tensor_ops.hpp>
+#include <miopen/implicitgemm_params.hpp>
+
+namespace miopen {
+namespace solver {
+
+PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm::PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm()
+    : PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm::
+          PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm(
+              64, 64, 2, 32, 32, 4, 16, 64, 16, false, false)
```
Comment (asroy) [inline src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp:45]: Please add comment here that GemmMFactor GemmNFactor, GemmKFactor are fixed value at this moment.

File: src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp [hunk 6]
```diff
@@ -0,0 +1,1106 @@
...
+
+void PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm::EuristicInit(const ConvolutionContext& ctx)
+{
+    PerformanceImplicitGemmWrwV4R4Xdlops_Padded_Gemm tmp;
+
+    // loop over certain ranges of tuning parameter
+    auto get_euristic_config = [&](auto is_valid_func) {
+        if(ctx.IsFp32())
+        {
+            tmp = {256, 256, 8, 128, 128, 4, 16, 64, 16, false, true};
```
Comment (asroy) [inline src/solver/conv_hip_implicit_gemm_wrw_v4r4_xdlops_padded_gemm.cpp:129]: Please add comment here that GemmMFactor GemmNFactor, GemmKFactor are fixed value at this moment.

**Review by asroy — CHANGES_REQUESTED** [review]

File: src/include/miopen/solver.hpp (hunk 1 above)
Comment (asleepzzz) [reply src/include/miopen/solver.hpp:1998]: if(GemmBThreadCopyMoreGemmKPack)
        {
            data_per_thread_copy_gemmk = gcd(GemmKPerBlock, tmp);
            data_per_thread_copy_gemmn = tmp / data_per_thread_copy_gemmk;
        }
        else
        {
            data_per_thread_copy_gemmn = gcd(GemmKPerBlock, tmp);
            data_per_thread_copy_gemmk = tmp / data_per_thread_copy_gemmn;
        }
non-padding kernel should also change name

**Review by asroy — DISMISSED** [review]

**Comment by asroy** [comment]
@asleepzzz formatting issue
http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/wrw_xdlops_padd/12/pipeline

**Comment by asleepzzz** [comment]
CI passed

**Review by asroy — APPROVED** [review]

**Pull request merged**